src/main/java/com/unstoppabledomains/resolution/ResolutionBuilder.java

```diff
@@ -0,0 +1,284 @@
+    }
+
+    /**
+     * @param providerUrl blockchain provider URL for ENS
+     * @return builder object to allow chaining
+     */
+    public ResolutionBuilder ensProviderUrl(String providerUrl) {
+        NSConfig nsConfig = serviceConfigs.get(NamingServiceType.ENS);
+        return this.providerUrl(nsConfig, providerUrl);
+    }
+
+    /**
+     * Applies the blockchain provider URL for UNS naming service 
+     * @param location the location of the UNS service (layer 1 or layer 2)
+     * @param providerUrl blockchain provider URL
+     * @return builder object to allow chaining
+     */
+    public ResolutionBuilder unsProviderUrl(UNSLocation location, String providerUrl) {
+        NSConfig nsConfig = unsConfigs.get(location);
+        return this.providerUrl(nsConfig, providerUrl);
+    }
+
+    private ResolutionBuilder providerUrl(NSConfig nsConfig, String providerUrl) {
+        Network chainId = getNetworkId(providerUrl);
+        if (chainId == null) {
+            chainId = nsConfig.getChainId();
+        }
+        nsConfig.setBlockchainProviderUrl(providerUrl);
+        nsConfig.setChainId(chainId);
+        return this;
+    }
+
+    /**
+     * @param contractAddress   address of `Registry` contract for ZNS
+     * @return builder object to allow chaining
+     */
+    public ResolutionBuilder znsContractAddress(String contractAddress) {
+        NSConfig nsConfig = serviceConfigs.get(NamingServiceType.ZNS);
+        nsConfig.setContractAddress(contractAddress);
+        return this;
+    }
+
+    /**
+     * @param contractAddress   address of `Registry` contract for ENS
+     * @return builder object to allow chaining
+     */
+    public ResolutionBuilder ensContractAddress(String contractAddress) {
+        NSConfig nsConfig = serviceConfigs.get(NamingServiceType.ENS);
+        nsConfig.setContractAddress(contractAddress);
+        return this;
+    }
+
+    /**
+     * Applies the address of `ProxyReader` contract for UNS naming service 
+     * @param location the location of the UNS service (layer 1 or layer 2)
+     * @param contractAddress   address of `ProxyReader` contract for UNS
+     * @return builder object to allow chaining
+     */
+    public ResolutionBuilder unsContractAddress(UNSLocation location, String contractAddress) {
+        NSConfig nsConfig = unsConfigs.get(location);
+        nsConfig.setContractAddress(contractAddress);
+        return this;
+    }
+
+    /**
+     * Configuration for <a href="https://infura.io">infura.io</a>
+     * blockchain provider with previously set network ID using
+     * {@link ResolutionBuilder#chainId(NamingServiceType, Network)} or default one
+     * ({@link Network#MAINNET})
+     *
+     * @param nsType    the naming service for which config is applied
+     * @param projectId Infura project ID
+     * @return builder object to allow chaining
+     */
+    public ResolutionBuilder infura(NamingServiceType nsType, String projectId) {
```

Comment:
@DancingAxolotl Actually I think infura() methods are redundant since user could set Infura endpoint using `*providerUrl()` methods. What do you think?

Reply:
I like it. Since we are changing builder methods anyway, it would be good to simplify it in the process.